lib/ecto/changeset/relation.ex

```diff
@@ -324,6 +325,14 @@ defmodule Ecto.Changeset.Relation do
     end
   end
 
+  defp key_as_int({key, val}) when is_binary(key) do
+    case Integer.parse(key) do
+      {key, _} -> {key, val}
```

Comment:
It should be rather:

```elixir
case Integer.parse(key) do
  {key, ""} -> {key, val}
  _ -> {key, val}
end
```

Otherwise we will convert strings starting with numbers and discard the rest of the string.